File: Scripts/Room/ShopRoomHandler.cs
using Roguelike.Data;
using RoguelikeMapGen;

namespace Roguelike.Logic.Handlers
{
    public class ShopRoomHandler : IRoomHandler
    {
        /// <summary>
        /// Generates a new shop inventory and transitions the game state to await player actions.
        /// </summary>
        public void Execute(GameRun run, Room room)
        {
            run.CurrentShop = new ShopInventory(run.CardPool, run.RelicPool, run.Rng);
            
            run.CurrentState = GameState.InShop;
        }

        /// <summary>
        /// Attempts to purchase a card from the shop. Called by the GameController.
        /// </summary>
        /// <returns>True if the purchase was successful, false otherwise.</returns>
        public static bool PurchaseCard(GameRun run, int cardIndex)
        {
            if (run.CurrentState != GameState.InShop || run.CurrentShop == null) return false;
            if (cardIndex < 0 || cardIndex >= run.CurrentShop.CardsForSale.Count) return false;

            var item = run.CurrentShop.CardsForSale[cardIndex];

            if (item.IsSold || run.TheHero.CurrentGold < item.Price)
            {
                return false;
            }

            run.TheHero.CurrentGold -= item.Price;
            run.TheHero.Deck.AddCardToMasterDeck(item.Item);
            item.IsSold = true;

            return true;
        }

        /// <summary>
        /// Attempts to purchase a relic from the shop. Called by the GameController.
        /// </summary>
        /// <returns>True if the purchase was successful, false otherwise.</returns>
        public static bool PurchaseRelic(GameRun run, int relicIndex)
        {
            if (run.CurrentState != GameState.InShop || run.CurrentShop == null) return false;
            if (relicIndex < 0 || relicIndex >= run.CurrentShop.RelicsForSale.Count) return false;

            var item = run.CurrentShop.RelicsForSale[relicIndex];

            if (item.IsSold || run.TheHero.CurrentGold < item.Price)
            {
                return false;
            }

            run.TheHero.CurrentGold -= item.Price;
            run.TheHero.Relics.Add(item.Item);
            item.IsSold = true;

            return true;
        }

        /// <summary>
        /// Cleans up shop state and returns to the map. Called by the GameController.
        /// </summary>
        public static void LeaveShop(GameRun run)
        {
            run.CurrentShop = null;
            run.CurrentState = GameState.OnMap;
        }
    }
}

File: python_roguelike/core/rooms/shop_room_handler.py
from .i_room_handler import IRoomHandler
from .shop_inventory import ShopInventory
from ...data.enums import GameState


class ShopRoomHandler(IRoomHandler):
    def execute(self, run, room):
        run.current_shop = ShopInventory(run.card_pool, run.relic_pool, run.rng)
        run.current_state = GameState.InShop

    @staticmethod
    def purchase_card(run, card_index: int) -> bool:
        if run.current_state != GameState.InShop or run.current_shop is None:
            return False
        if card_index < 0 or card_index >= len(run.current_shop.cards_for_sale):
            return False

        item = run.current_shop.cards_for_sale[card_index]
        if item.is_sold or run.the_hero.current_gold < item.price:
            return False

        run.the_hero.current_gold -= item.price
        run.the_hero.deck.add_card_to_master_deck(item.item)
        item.is_sold = True
        return True

    @staticmethod
    def purchase_relic(run, relic_index: int) -> bool:
        if run.current_state != GameState.InShop or run.current_shop is None:
            return False
        if relic_index < 0 or relic_index >= len(run.current_shop.relics_for_sale):
            return False

        item = run.current_shop.relics_for_sale[relic_index]
        if item.is_sold or run.the_hero.current_gold < item.price:
            return False

        run.the_hero.current_gold -= item.price
        run.the_hero.relics.append(item.item)
        item.is_sold = True
        return True

    @staticmethod
    def leave_shop(run):
        if run.current_state != GameState.InShop:
            return
        run.current_shop = None
        run.current_state = GameState.OnMap
